uc-crux-llvm/src/UCCrux/LLVM/Overrides/Check.hs

```diff
@@ -0,0 +1,358 @@
+import           Lang.Crucible.LLVM.Intrinsics (LLVM, LLVMOverride(..), basic_llvm_override)
+
+-- crux-llvm
+import           Crux.LLVM.Overrides (ArchOk)
+
+-- uc-crux-llvm
+import           UCCrux.LLVM.Constraints (Constraint, Constraints, ConstrainedShape(..), argConstraints, globalConstraints, ppConstraints)
+import           UCCrux.LLVM.Context.App (AppContext, log)
+import           UCCrux.LLVM.Context.Module (ModuleContext, moduleDecls, moduleTypes)
+import qualified UCCrux.LLVM.Errors.Unimplemented as Unimplemented
+import           UCCrux.LLVM.FullType.CrucibleType (SomeIndex(SomeIndex), translateIndex, toCrucibleType)
+import           UCCrux.LLVM.FullType.StorageType (toStorageType)
+import           UCCrux.LLVM.FullType.Type (FullType(FTPtr), FullTypeRepr, MapToCrucibleType, ToCrucibleType, pointedToType, arrayElementType)
+import           UCCrux.LLVM.Logging (Verbosity(Hi))
+import           UCCrux.LLVM.Module (FuncSymbol, funcSymbol)
+import           UCCrux.LLVM.Overrides.Polymorphic (PolymorphicLLVMOverride, makePolymorphicLLVMOverride)
+import           UCCrux.LLVM.Run.Result (BugfindingResult)
+import qualified UCCrux.LLVM.Run.Result as Result
+import           UCCrux.LLVM.Setup.Constraints (constraintToPred)
+import qualified UCCrux.LLVM.Shape as Shape
+{- ORMOLU_ENABLE -}
+
+newtype CheckOverrideName = CheckOverrideName {getCheckOverrideName :: Text}
+  deriving (Eq, Ord, Show)
+
+declName :: L.Declare -> Text
+declName decl =
+  let L.Symbol name = L.decName decl
+   in Text.pack name
+
+-- TODO: Alignment...?
+doLoad ::
+  forall arch m sym ft.
+  IsSymInterface sym =>
+  HasLLVMAnn sym =>
+  ArchOk arch =>
+  (?memOpts :: LLVMMem.MemOptions) =>
+  ModuleContext m arch ->
+  sym ->
+  LLVMMem.MemImpl sym ->
+  Crucible.RegValue sym (ToCrucibleType arch ('FTPtr ft)) ->
+  FullTypeRepr m ('FTPtr ft) ->
+  IO (Pred sym, Maybe (Crucible.RegValue sym (ToCrucibleType arch ft)))
+doLoad modCtx sym mem val fullTypeRepr =
+  do let pointedToRepr = pointedToType (modCtx ^. moduleTypes) fullTypeRepr
+         typeRepr = toCrucibleType modCtx pointedToRepr
+     -- TODO Is this alignment right?
+     partVal <-
+       LLVMMem.loadRaw sym mem val (toStorageType pointedToRepr) noAlignment
+     case partVal of
+       LLVMMem.Err p -> return (p, Nothing)
+       LLVMMem.NoErr p ptdToVal ->
+         (p,) . Just <$> LLVMMem.unpackMemValue sym typeRepr ptdToVal
+
+
+ifoldMapM ::
+  FoldableWithIndex i t =>
+  Monoid m =>
+  Monad f =>
+  (i -> a -> f m) ->
+  t a ->
+  f m
+ifoldMapM f = ifoldrM (\i x acc -> fmap (<> acc) (f i x)) mempty
+
+ifoldMapMFC ::
+  FoldableFCWithIndex t =>
+  Monoid m =>
+  Monad g =>
+  (forall x. IndexF (t f z) x -> f x -> g m) ->
+  t f z ->
+  g m
+ifoldMapMFC f = ifoldrMFC (\i x acc -> fmap (<> acc) (f i x)) mempty
+
+-- | Create a predicate that checks that a Crucible(-LLVM) value conforms to the
+-- 'ConstrainedShape'.
+--
+-- TODO: Add selector for provenance
```

Comment:
Could we add a note here about why the return value is a `Seq (Pred sym)` instead of just a `Pred sym`?

Reply:
Yeah, good call - In brief, it's because eventually I'll add more "provenance" information, e.g. what `Constraint` is this `Pred` from, and *where* in the value does that constraint apply (i.e., a `Cursor`).

Reply:
I added a new commit that just resolves the TODO and takes care of this along the way.